nsfc-plugin/src/main/java/org/osc/controller/nsfc/api/NeutronSfcSdnRedirectionApi.java

```diff
@@ -147,45 +366,61 @@ public NetworkElement registerNetworkElement(List<NetworkElement> portPairGroupL
+        checkArgument(portChain != null && portChain.getId() != null,
+                      "Cannot find %s by id: %s!", "Service Function Chain", serviceFunctionChain.getElementId());
 
         portChain = Builders.portChain().from(portChain)
                             .portPairGroups(Collections.emptyList()).build();
-        this.osClient.sfc().portchains().update(portChain.getId(), portChain);
+        this.osCalls.updatePortChain(portChain.getId(), portChain);
         this.utils.validatePPGList(portPairGroupList);
 
         List<String> portPairGroupIds = portPairGroupList
                 .stream()
                 .map(ppg -> ppg.getElementId())
-                .collect(Collectors.toList());
+                .collect(toList());
+
+        portChain = Builders.portChain().portPairGroups(portPairGroupIds).build();
+        PortChain portChainUpdated = this.osCalls.updatePortChain(serviceFunctionChain.getElementId(), portChain);
 
-        portChain = Builders.portChain().from(portChain).portPairGroups(portPairGroupIds).build();
-        PortChain portChainUpdated = this.osClient.sfc().portchains().update(portChain.getId(), portChain);
-        return new NetworkElementEntity(portChainUpdated.getId(),
-                                        new ArrayList<>(), new ArrayList<>(),
-                                        portPairGroupList.get(0).getParentId());
+        List<PortPairGroupEntity> portPairGroups =
+                portPairGroupIds.stream().map(id -> new PortPairGroupEntity(id)).collect(toList());
+        ServiceFunctionChainEntity retVal = new ServiceFunctionChainEntity(portChainUpdated.getId());
+        portPairGroups.stream().forEach(p -> p.setServiceFunctionChain(retVal));
+        retVal.setPortPairGroups(portPairGroups);
+        return retVal;
     }
 
     @Override
     public void deleteNetworkElement(NetworkElement serviceFunctionChain) throws Exception {
-        this.utils.throwExceptionIfNullElementAndId(serviceFunctionChain, "Service Function Chain Id");
-        PortChain portChain = this.osClient.sfc().portchains().get(serviceFunctionChain.getElementId());
-        this.utils.throwExceptionIfCannotFindById(portChain,"Service Function Chain", serviceFunctionChain.getElementId());
-        ActionResponse response = this.osClient.sfc().portchains().delete(serviceFunctionChain.getElementId());
+        checkArgument(serviceFunctionChain != null && serviceFunctionChain.getElementId() != null,
+                      "null passed for %s !", "Service Function Chain Id");
+
+        PortChain portChain = this.osCalls.getPortChain(serviceFunctionChain.getElementId());
+
+        checkArgument(portChain != null && portChain.getId() != null,
+                      "Cannot find %s by id: %s!", "Service Function Chain", serviceFunctionChain.getElementId());
+
+        ActionResponse response = this.osCalls.deletePortChain(serviceFunctionChain.getElementId());
 
         if (!response.isSuccess()) {
```

Comment:
if we get 404 we should not throw exception. Same in other places. maybe log a warning in 404 case.